velox/functions/prestosql/MapFromEntries.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <iostream>
+
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/expression/VectorReaders.h"
+#include "velox/functions/lib/RowsTranslationUtil.h"
+
+namespace facebook::velox::functions {
+namespace {
+// See documentation at https://prestodb.io/docs/current/functions/map.html
+class MapFromEntriesFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 1);
+    auto& arg = args[0];
+
+    VectorPtr localResult;
+
+    // Input can be constant or flat.
+    if (arg->isConstantEncoding()) {
+      auto* constantArray = arg->as<ConstantVector<ComplexType>>();
+      const auto& flatArray = constantArray->valueVector();
+      const auto flatIndex = constantArray->index();
+
+      SelectivityVector singleRow(flatIndex + 1, false);
+      singleRow.setValid(flatIndex, true);
+      singleRow.updateBounds();
+
+      localResult = applyFlat(singleRow, flatArray.get(), outputType, context);
+      localResult =
+          BaseVector::wrapInConstant(rows.size(), flatIndex, localResult);
+    } else {
+      exec::DecodedArgs decodedArgs(rows, args, context);
```

Comment:
Thank you!!! Your sketch is super helpful for me to decode such a nested complex vector.  

FYI: I know the gist of different simple vector layouts when I was in Saber, but I am not super familiar with the surrounding helper / wrapper classes and the usage scenarios, like SelectivityVector, complex vector layouts (this is not contributed from Saber), DecodedVector (I wrote our own vector SerDe classes when I was in Saber). So excited to witness so many new velox features since then, and happy to refresh my arsenal.

Also thanks for your quick workaround for bug #3424 and #3533, I amended a comment - it actually affects a few map functions.

Reply:
> I modified the Fuzzer to workaround https://github.com/facebookincubator/velox/issues/3533 and also modified DecodedVector to workaround a bug that @kagamiori Wei is looking into (https://github.com/facebookincubator/velox/pull/3424)

Since Wei is working on the fuzzer bugs, I will leave the fuzzer code unchanged with CI failures. Once these two bugs are fixed, I will rebase from main and land this PR. Does it sound good?

Reply:
> Since Wei is working on the fuzzer bugs, I will leave the fuzzer code unchanged with CI failures. Once these two bugs are fixed, I will rebase from main and land this PR. Does it sound good?

Indeed, this PR has a dependency on fixing Fuzzer and DecodedVector. Just to clarify, after rebasing the PR, you'll need to re-run Fuzzer to make sure there are not failures and the PR will need to be reviewed before it is ready for landing.

Reply:
@mbasmanova - regarding the fuzzer failure, I did some debugging and found that the exception is thrown at this line:
```c++
    VectorPtr wrappedKeys;
    VectorPtr wrappedValues;
    if (decodedRow->isIdentityMapping()) {
      wrappedKeys = rowVector->childAt(0);
      wrappedValues = rowVector->childAt(1);
    } else {
      wrappedKeys = decodedRow->wrap(
          rowVector->childAt(0), *inputRowVector, inputRowVector->size()); // <---- THROW EXCEPTION: 
          // "Dictionary index must be less than base vector's size. Index:..." 
          // mostly reproduced at the simplified path + the duplicate-key case
      wrappedValues = decodedRow->wrap(
          rowVector->childAt(1), *inputRowVector, inputRowVector->size());
    }
```
Question:
Why do you call `decodedRow->wrap()` to wrap the keys and values vectors for non-flat encodings in your [sketch code](https://github.com/facebookincubator/velox/compare/main...mbasmanova:velox-1:map_from_entries_simpleudf?expand=1)? Maybe I'm wrong, but I feel we should compute the total size, new indices, and call `BaseVector::wrapInDictionary()` to wrap the keys and values. Refer to [map()'s implementation](https://github.com/facebookincubator/velox/blob/22d9efb0645801aa2f193740827829c088210884/velox/functions/prestosql/Map.cpp#L163-L173). What do you think?

I noticed that this edge case usually happens when keys is a nested dictionary encoding or even more complex encodings.

Reply:
Found the root cause, see https://github.com/facebookincubator/velox/issues/3687